pkg/auth/cache/synchronized_access_cache.go

```diff
@@ -179,3 +185,21 @@ func (s *SynchronizedAccessCache) Start(ctx context.Context) {
 		}()
 	})
 }
+
+func (s *SynchronizedAccessCache) List(ctx context.Context, subject rbacv1.Subject) []corev1.Namespace {
+	namespaces := s.cache.List(subject)
+	namespaceList := make([]corev1.Namespace, 0, len(namespaces))
+
+	for _, ns := range namespaces {
+		namespace := corev1.Namespace{}
+		err := s.namespaceLister.Get(ctx, types.NamespacedName{Name: ns}, &namespace)
```

Comment:
this way we are getting access from the `AccessCache` and namespaces from the `ResourcesCache`. This is coupling the `List` operation to the two caches.

It may happen that the namespaces in the `ResourcesCache` were updated before the access in the `AccessCache` and this may introduce some errors.
I'd suggest to cache the namespaces used during the synch operation in the `AccessCache`, so that we always provide consistent results.

Reply:
Also if we cache it in the AccessCache, we can implement a slightly more performant version of this algorithm by sorting namespaces ahead-of-time and using binary search just-in-time